### DISCUSSION TITLE
How to configure PostgreSQL search_path for Prisma

### DISCUSSION THREAD

**Discussion by oskarrough** [prisma/prisma#14529] [Q&A] [answered]
Hello, we have a PostgreSQL database with two schemas:

- `public`
- `playground`

In the `playground` schema there is a `post` table. I can query it with `select * from playground.post` OR `select * from post`, because the `search_path` is set to `public, playground`.

Our `.env` file: `DATABASE_URL="postgres://user:pass@host:5432/db?schema=playground"`.

I run `prisma db pull` and it correctly inspect and writes the prisma.schema. And now comes the issue, running `await prisma.post.findMany()` returns an error:

> The table `public.post` does not exist in the current database.

… which is true, because the `post` table is in the playground schema. How do I tell Prisma client to look there as well (or only)?

**Comment by austincrim** [comment, marked as answer]
It's not yet possible to query across multiple schemas (without using [`$queryRaw`](https://www.prisma.io/docs/concepts/components/prisma-client/raw-database-access#queryraw)), but it's something we are actively looking at.

Feel free to read the latest updates on [this issue](https://github.com/prisma/prisma/issues/1122) and chime in with any feedback!

**Reply by oskarrough** [reply]
I did see that issue, but in this scenario only a single SQL schema needs to be queries. It's just not called `public`. Is that supported without `queryRaw`? In any case, thank you 👍

**Reply by austincrim** [reply]
If you only want to query the `playground` schema, that it possible using Prisma, even without `$queryRaw`.

Your connection string looks correct to me, have you run `npx prisma generate`?

**Reply by oskarrough** [reply]
Yes I did.

    prisma.$queryRaw`select * from public.post` // does not work (and should not work)
    prisma.$queryRaw`select * from playground.post` // works
    prisma.$queryRaw`select * from post` // works
    prisma.post.findMany()` // still throws an error that `public.post` table doesn't exist
    
Is the `DATABASE_URL` connection string the only place I have to define to SQL schema? I don't see any mention of my "playground" schema inside schema.prisma.

After running `npx prisma generate`, inside `node_modules/.prisma` there is no mention of the custom schema either. Is that correct?. Seems the schema defined in the connection string is initially used for `npx prisma db pull` but otherwise ignored, in my case.

**Reply by janpio** [reply]
The schema is only defined in the connection string, similar to the database name. If you set it there, it should impact all queries. Any local search_path settings should not really matter in this case.

Do other models/tables work fine?
Just to make 100% sure: You are using the env var for `url` in your `datasource` configuration in the schema?

**Reply by oskarrough** [reply]
Without knowing how prisma really works, I believe the issue is that the schema from `DATABASE_URL` inside `.env` isn't being used for the Prisma client's queries. It prepends all my tables with the `public.` schema at the moment, not sure why.

If it would either a) not specify a schema b) use the schema from `env("DATABASE_URL")`, it should work. I'm wondering if it has to do with the `?schema=` part of the connection string and my PostgreSQL being on Heroku?

Here's a preview of my `schema.prisma`:

```prisma
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model user {
  createddate    DateTime? @db.Timestamp(6)
  name           String?   @db.VarChar(121)
  systemmodstamp DateTime? @db.Timestamp(6)
  email          String?   @db.VarChar(128)
  sfid           String?   @unique(map: "hcu_idx_user_sfid") @db.VarChar(18)
  id             Int       @id @default(autoincrement())
  alias          String?   @db.VarChar(8)

  @@index([systemmodstamp], map: "hc_idx_user_systemmodstamp")
}
```

**Reply by janpio** [reply]
If it is reading the connection string from the env var, then it is also using that `schema` as the default search path. It only uses `public` when no `schema` is supplied.

Is your connection string really valid? Can you shared the part starting from the last `/`?

**Reply by oskarrough** [reply]
@janpio thanks. I assume it is correct, as the `prisma db pull` works fine. This is how it looks:

`DATABASE_URL="postgres://a:b@c:5432/d?schema=salesforce"`

**Reply by janpio** [reply]
That looks as expected.

In theory Prisma Client should load the connection string the same way as `db pull`. It is even weirded, as some queries seem to work.

**Reply by oskarrough** [reply]
With the help of a friend, we finally found it. The issue was a confusion with the environment variables.

- Prisma's getting started documentation tells to add a `DATABASE_URL` with schema to `.env`.
- Next.js, which we are using, uses `.env.local` locally

So we have two variables of the same name.

- `prism db pull` looks in `.env`
- the Prisma client inside next.js looks in `.env.local` 

And the .env.local didn't contain the schema........ Sorry about this. But at least we figured it out, and the issue was NOT Heroku Connect or Prisma.

Can you tell `./prisma/prisma.schema` from which file to load the env variables?

**Reply by austincrim** [reply]
This looks to be related to [how Next.js loads environment variables](https://nextjs.org/docs/basic-features/environment-variables#environment-variable-load-order). It looks for `.env.local` first and stops when it finds it.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
